### PR TITLE
Enable more MongoDB Micrometer metrics

### PR DESCRIPTION
**Pull request by vkn**
Currently only connection pool micrometer metrics are supported. This PR enables all available metrics based on `io.micrometer.core.instrument.binder.mongodb.MongoMetricsCommandListener`

All unit tests in mongodb-client extension are based on microprofile metrics, so it was not possible to add tests for micrometer, without modifying dependencies and existing tests. However I've added an assertion in MongodbPanacheResourceTest.java in integration-tests/mongodb-panache  module

### CODE REVIEW

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java [hunk 1]
```diff
@@ -124,6 +124,17 @@ AdditionalIndexedClassesBuildItem includeMongoCommandListener(MongoClientBuildTi
...
 
+    @BuildStep
+    AdditionalIndexedClassesBuildItem includeMongoCommandMetricListener(
+            MongoClientBuildTimeConfig buildTimeConfig,
+            Optional<MetricsCapabilityBuildItem> metricsCapability) {
+        return metricsCapability
+                .filter(cap -> buildTimeConfig.metricsEnabled && cap.metricsSupported(MetricsFactory.MICROMETER))
+                .map(supported -> new AdditionalIndexedClassesBuildItem(
+                        MongoClientRecorder.getMicrometerCommandListenerClassName()))
+                .orElseGet(AdditionalIndexedClassesBuildItem::new);
```
Comment (gsmet) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:135]: Any chance you could implement this with simple ifs (with early returns for edge cases)?

Because this is a bit hard to follow and I'm worried we will regret it when fixing a bug one early morning without coffee :).

**Review by gsmet — COMMENTED** [review]
Thanks for the PR! I added a small comment inline.

I will defer to Loïc and Bruno to check the feature.

**Review by brunobat — CHANGES_REQUESTED** [review]
It looks ok to me but could you please add a test entry here: https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java#L24

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java (hunk 1 above)
Comment (vkn) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java]: @gsmet pushforced the change

**Comment by vkn** [comment]
> It looks ok to me but could you please add a test entry here:
> 
> https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java#L24

@brunobat the test is based on microprofile metrics. to add test for the micrometer metrics, we would have to refactor tests and create separate tests for micrometer and microprofile metrics. or do you see an easier way

**Comment by brunobat** [comment]
> > It looks ok to me but could you please add a test entry here:
> > https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java#L24
> 
> @brunobat the test is based on microprofile metrics. to add test for the micrometer metrics, we would have to refactor tests and create separate tests for micrometer and microprofile metrics. or do you see an easier way

yeah, sorry, didn't look at the imports... Then It's worse than I thought.
Can you please add a test class for Micrometer under the deployment module?

**Comment by vkn** [comment]
> > > It looks ok to me but could you please add a test entry here:
> > > https://github.com/quarkusio/quarkus/blob/ebee0a4b66d1b4dba9c51165e78d22fbf3b90d3e/extensions/mongodb-client/deployment/src/test/java/io/quarkus/mongodb/MongoMetricsTest.java#L24
> > 
> > 
> > @brunobat the test is based on microprofile metrics. to add test for the micrometer metrics, we would have to refactor tests and create separate tests for micrometer and microprofile metrics. or do you see an easier way
> 
> yeah, sorry, didn't look at the imports... Then It's worse than I thought. Can you please add a test class for Micrometer under the deployment module?

@brunobat The deployment module pom has `quarkus-smallrye-metrics-deployment` dependency. so microprofile metrics are on the class path and there is no way to add micrometer metrics, unless you agree to remove it from the pom xml and add the dependency in QuarkusUnitTest programmatically. The existing test have to be then adjusted and not import microprofile metrics, as far as I can tell

**Comment by loicmathieu** [comment]
>  The deployment module pom has quarkus-smallrye-metrics-deployment dependency. so microprofile metrics are on the class path and there is no way to add micrometer metrics,

Mongodb metrics support preclude our switch to Micrometer by default that's why the test is based on Microprofile metrics, I'm OK to switch to test to Micrometer.

@vkn thanks a lot for your contribution. MongoDB could emit a lot of commands, I wonder if this listener should be added always of if we could be able to disable it for performance reason.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java [hunk 2]
```diff
@@ -99,6 +100,10 @@ public ConnectionPoolListener get() {
         };
     }
 
+    public static String getMicrometerCommandListenerClassName() {
+        return MicrometerCommandListener.class.getName();
+    }
+
```
Comment (loicmathieu) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java:106]: As this is used in a single place you can directly add the class name at the use site.
This would also avoid loading the class at runtime if not used wich would be better.

**Review by loicmathieu — DISMISSED** [review]
As other already said, a test would be great

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java (hunk 2 above)
Comment (vkn) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java:106]: > As this is used in a single place you can directly add the class name at the use site. This would also avoid loading the class at runtime if not used wich would be better.

micrometer wasn't on class path in deployment module, that's why I've put it there. I'll add optional dependency and move the code to the deployment

**Comment by vkn** [comment]
@brunobat @loicmathieu 
I've changed the tests to be based on micrometer. `MongoMetricsTest ` verifies the new `mongodb.driver.commands` metrics are available. Please have a look

File: extensions/mongodb-client/deployment/pom.xml [hunk 3]
```diff
@@ -63,8 +63,8 @@
         </dependency>
         <dependency>
             <groupId>io.quarkus</groupId>
-            <artifactId>quarkus-smallrye-metrics-deployment</artifactId>
-            <scope>test</scope>
+            <artifactId>quarkus-micrometer-registry-prometheus-deployment</artifactId>
```
Comment (brunobat) [inline extensions/mongodb-client/deployment/pom.xml:66]: Can you make it with just the `quarkus-micrometer` extension?

Comment (vkn) [reply extensions/mongodb-client/deployment/pom.xml:66]: @brunobat I get:
```
[ERROR] Rule 4: io.quarkus.enforcer.BansRuntimeDependency failed with message:
[ERROR] 1 illegal runtime dependencies found that have to be replaced with their -deployment counterparts:
[ERROR]         quarkus-micrometer
```

Comment (brunobat) [reply extensions/mongodb-client/deployment/pom.xml:66]: you need the `quarkus-micrometer-deployment`

Comment (brunobat) [reply extensions/mongodb-client/deployment/pom.xml]: I mean, replace `quarkus-micrometer-registry-prometheus-deployment` with `quarkus-micrometer-deployment`.
I'm not sure if it will cause you an error on the tests, if so `quarkus-micrometer-registry-prometheus-deployment` can be there but as a test dependency.
The reason is that the `quarkus-micrometer-registry-prometheus-deployment` is a particular telemetry output lib for prometheus, there are other implementations like otlp and we should avoid depending directly on them o the main code.

Comment (vkn) [reply extensions/mongodb-client/deployment/pom.xml]: Makes sense. I did exactly that. Still need to fix imports sorting. Then it should be ok

Comment (brunobat) [reply extensions/mongodb-client/deployment/pom.xml]: Thanks!

**Review by brunobat — APPROVED** [review]
Looks good to me.
Thanks @vkn !

**Review by gsmet — CHANGES_REQUESTED** [review]
Thanks for the adjustment. I requested a small change as we shouldn't produce an empty `AdditionalIndexedClassesBuildItem` when we don't want to produce anything.

Once this is fixed, it's all good to go.

I can do it if you prefer but I thought you might be interested in doing it yourself. Let me know!

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java [hunk 4]
```diff
@@ -124,6 +125,21 @@ AdditionalIndexedClassesBuildItem includeMongoCommandListener(MongoClientBuildTi
         return new AdditionalIndexedClassesBuildItem();
     }
 
+    @BuildStep
+    AdditionalIndexedClassesBuildItem includeMongoCommandMetricListener(
+            MongoClientBuildTimeConfig buildTimeConfig,
+            Optional<MetricsCapabilityBuildItem> metricsCapability) {
+        if (!buildTimeConfig.metricsEnabled) {
+            return new AdditionalIndexedClassesBuildItem();
```
[the remark this answers is not in the document]
Comment (vkn) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java]: @gsmet  I've adjusted the code as you requested. Thanks!

**Review by gsmet — APPROVED** [review]
